### PR TITLE
DX: Resolve/Ignore PHPStan issues on level 6 + bump to level 7 with new baseline

### PR DESCRIPTION
**Pull request by Wirone**
I've fixed issues from level 6 where it made sense. Some of the errors were ignored, as we purposefully violate contravariance in some places, we also re-use variables in multiple loops across methods so we need to treat loops like part of the method scope, not separate scope (more info [here](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/7969#discussion_r1581955134)). I've added explicit paths for some ignore rules, so we can avoid such code in other places (it's easier to write new code that will follow rules than refactor existing one).

### CODE REVIEW

File: src/Console/ConfigurationResolver.php [hunk 1]
```diff
@@ -554,7 +554,7 @@ private function computeConfigFiles(): array
             $configDir = $path[0];
         } else {
             $dirName = pathinfo($path[0], PATHINFO_DIRNAME);
-            $configDir = $dirName ?: $path[0];
+            $configDir = is_dir($dirName) ? $dirName : $path[0];
```
Comment (keradus) [inline src/Console/ConfigurationResolver.php:557]: we already took PATHINFO_DIRNAME line above, why we need to check if returned dirname is dir ?

File: src/Fixer/Alias/NoMixedEchoPrintFixer.php [hunk 2]
```diff
@@ -80,12 +80,15 @@ public function isCandidate(Tokens $tokens): bool
...
+     * @uses fixPrintToEcho()
+     * @uses fixEchoToPrint()
+     */
     protected function applyFix(\SplFileInfo $file, Tokens $tokens): void
     {
-        $callBack = $this->callBack;
         foreach ($tokens as $index => $token) {
             if ($token->isGivenKind($this->candidateTokenType)) {
-                $this->{$callBack}($tokens, $index);
+                \call_user_func_array($this->callBack, [$tokens, $index]);
```
Comment (keradus) [inline src/Fixer/Alias/NoMixedEchoPrintFixer.php:91]: if ($this->candidateTokenType === T_PRINT) {
    $this->fixPrintToEcho($tokens, $index);
} else {
    $this->fixEchoToPrint($tokens, $index);
}

no need to have generic mechanism for any callback if we have only 2 callbacks.
and with this, no need to call method by it's string name

File: src/Console/ConfigurationResolver.php (hunk 1 above)
Comment (Wirone) [reply src/Console/ConfigurationResolver.php:557]:
```php
$paths = [__FILE__, '/non/existing/path/foo.php'];

foreach ($paths as $path) {
    $dirName = pathinfo($path, PATHINFO_DIRNAME);
    $configDir = is_dir($dirName) ? $dirName : $path;

    var_dump($dirName, is_dir($dirName), $configDir);
}
```

```shell
$ php /Volumes/Projects/~Github/PHP-CS-Fixer/PHP-CS-Fixer/reproducer.php

string(51) "/Volumes/Projects/~Github/PHP-CS-Fixer/PHP-CS-Fixer"
bool(true)
string(51) "/Volumes/Projects/~Github/PHP-CS-Fixer/PHP-CS-Fixer"
string(18) "/non/existing/path"
bool(false)
string(26) "/non/existing/path/foo.php"
```

`PATHINFO_DIRNAME` only determines which part is parsed and returned, it does not enforce validity. With `$dirName ?: $path[0]` PHPStan reports `Short ternary operator is not allowed. Use null coalesce operator if applicable or consider using long ternary.`. Changing it to `$dirName ? $dirName : $path[0]` leads to `Only booleans are allowed in a ternary operator condition, string given`, hence the explicit check with `is_dir()`.

Comment (keradus) [reply src/Console/ConfigurationResolver.php:557]: thank you :thanks:

File: phpstan.dist.neon [hunk 3]
```diff
@@ -5,22 +5,56 @@ includes:
...
 
 parameters:
-    level: 6
+    level: 7
     paths:
         - src
         - tests
     excludePaths:
         - tests/Fixtures
+    polluteScopeWithLoopInitialAssignments: true # Do not enforce assignments outside of the loops
```
Comment (keradus) [inline phpstan.dist.neon:14]: OKish 😅

File: phpstan.dist.neon [hunk 4]
```diff
@@ -5,22 +5,56 @@ includes:
...
+                - src/Fixer/Whitespace/StatementIndentationFixer.php
+                - tests/DocBlock/TypeExpressionTest.php
+        -
+            message: '/^Foreach overwrites \$index with its key variable.$/'
+            paths:
+                - src/Fixer/DoctrineAnnotation/DoctrineAnnotationIndentationFixer.php
+                - src/Fixer/Import/GlobalNamespaceImportFixer.php
+                - src/Fixer/LanguageConstruct/GetClassToClassKeywordFixer.php
+
+        # Types related to tokens collections are purposefully narrowed, because we expect only tokens there
```
Comment (keradus) [inline phpstan.dist.neon:42]: i wonder if we can instruct on class level that item is given type and not `mixed`

File: phpstan.dist.neon [hunk 5]
```diff
@@ -5,22 +5,56 @@ includes:
...
+        - '/^Parameter .* of method PhpCsFixer\\Tokenizer\\Tokens::offsetSet\(\) should be contravariant with parameter .* of method .*::offsetSet\(\)$/'
+        - '/^Parameter .* of method PhpCsFixer\\Tests\\Test\\TokensWithObservedTransformers::offsetSet\(\) should be contravariant with parameter .* of method .*::offsetSet\(\)$/'
+
+        # This one is tricky, because adding @param to PHPDoc causes failures on PHP7.4 with lowest deps (phpdoc_to_param_type), using @phpstan-param causes failures on all PHP versions (phpdoc_add_missing_param_annotation)
+        - '/^Parameter #1 \$class \(string\) of method PhpCsFixer\\StdinFileInfo::get(File|Path)+Info+\(\) should be contravariant with parameter \$class \(string\|null\) of method SplFileInfo::get(File|Path)+Info\(\)$/'
+
+        # Contract for configurable fixers defines `array<string, mixed>`, but we want narrowed array shape for each fixer
```
Comment (keradus) [inline phpstan.dist.neon:50]: idea: is it possible to define on base class/interface the templated contract on `config=T / array{...}` and `T` is defined on rule level?

Comment (keradus) [reply phpstan.dist.neon:50]: ok, i see the link

File: src/Fixer/ArrayNotation/ArraySyntaxFixer.php [hunk 6]
```diff
@@ -82,13 +82,15 @@ public function isCandidate(Tokens $tokens): bool
...
+     * @uses fixToLongArraySyntax()
+     */
     protected function applyFix(\SplFileInfo $file, Tokens $tokens): void
     {
-        $callback = $this->fixCallback;
-
         for ($index = $tokens->count() - 1; 0 <= $index; --$index) {
             if ($tokens[$index]->isGivenKind($this->candidateTokenKind)) {
-                $this->{$callback}($tokens, $index);
+                \call_user_func_array($this->fixCallback, [$tokens, $index]);
```
[the remark this answers is not in the document]
Comment (keradus) [reply src/Fixer/ArrayNotation/ArraySyntaxFixer.php:93]: and later.

but also OK to call them exceptions for now

[the remark this answers is not in the document]
Comment (mvorisek) [reply src/Fixer/ArrayNotation/ArraySyntaxFixer.php:93]: ($this->fixCallback)($tokens, $index);

[the remark this answers is not in the document]
Comment (mvorisek) [reply src/Fixer/ArrayNotation/ArraySyntaxFixer.php:93]:
```suggestion
                ($this->fixCallback)($tokens, $index);
```

File: phpstan.dist.neon (hunk 3 above)
Comment (Wirone) [reply phpstan.dist.neon:14]: It's basically something we rely on in our code, as we often have loops where we iterate `$index`, change it within loop, and then operate on its value outside of the loop. We would need to refactor a lot of stuff to threat loops as separate scope.

File: phpstan.dist.neon [hunk 7]
```diff
@@ -5,22 +5,56 @@ includes:
     - dev-tools/phpstan/baseline.php
 
 parameters:
-    level: 6
```
[the remark this answers is not in the document]
Comment (Wirone) [reply phpstan.dist.neon:8]: I would like to merge this first, and then rebase #7969 and see what's the result. Even if we would keep constraints bump only, then let's merge it next.

File: src/Fixer/ArrayNotation/ArraySyntaxFixer.php (hunk 6 above)
[the remark this answers is not in the document]
Comment (Wirone) [reply src/Fixer/ArrayNotation/ArraySyntaxFixer.php]: Even though @mvorisek's suggestion works, I chose to get rid of callbacks stored in private property. Having explicit if/else and calling methods directly allowed to remove callback property, `configure()` method, `@uses` annotations etc, the code is just cleaner.

**Pull request merged**